lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -4721,12 +4721,22 @@ ConstraintResult GenericSignatureBuilder::addTypeRequirement(
         // type that makes sense to use here, but, in practice, all
         // getLookupConformanceFns used in here don't use that parameter anyway.
         auto dependentType = CanType();
-        auto conformance =
-            getLookupConformanceFn()(dependentType, subjectType, proto->getDecl());
 
-        // FIXME: diagnose if there's no conformance.
-        if (conformance) {
-          addConditionalRequirements(*this, *conformance, inferForModule);
+        auto D = getGenericParams().front()->getDecl();
+        auto DC = D ? D->getDeclContext() : nullptr;
+        auto NTD = DC ? DC->getAsNominalTypeOrNominalTypeExtensionContext()
+                      : nullptr;
+        auto subjectNTD = subjectType->getAnyNominal();
+
+        if (!NTD || !subjectNTD ||
+            !NTD->getDeclaredType()->isEqual(subjectNTD->getDeclaredType())) {
```

Comment:
I think one of us isn't understanding the other: I agree that we can't just compare `NTD->getDeclaredType()` with `subjectType`, but I wasn't proposing that. I believe that the declared types of two decls are equal if and only if the decls are equal (i.e. are the exactly the same decl), meaning this line can just be `NTD != subjectNTD`, and there's no need to do the work to get the declared types.

Reply:
Will this work if we are comparing `Foo<T> and `Foo<Int>`? Example:
```
extension Foo: P where Foo<Int>: P {...}
```

Reply:
Yes; getting down to the `NominalTypeDecl`s gets away from all the information about a specific instantiation of the type. Each type declaration corresponds to a single (nominal type) `Decl`, and one way to get a type corresponding to an instance of this unique `Decl` is `getDeclaredType`.

The problem here is we've got lots of `Type`s that all might connect to the same real underlying type, e.g., as you say, `Foo<T>` and `Foo<Int>`, and we want to make sure those match because they're both instantiations of the `Foo` type. The code is currently mapping both of those to a decl, and then asking for "what's the type that this decl declares" on each to compare, I'm suggesting we just map to the decl and compare that: if we have `Foo<T>` and `Bar<Int>` the nominal type decls will be different (the one for `Foo` and the one for `Bar`), and if we have two `Foo`s they'll both be the decl for `Foo`.

Reply:
Thank you! I was confusing types with type declarations, unbelievable 😅